### ISSUE TITLE
Problem in parallel training

### ISSUE THREAD

**Issue by walkerning** [walkerning/nics_fix_pytorch#2]
Currently, when doing parallel training across multiple GPUs (e.g. DataParallel), the fixed-point config for every module is unprotectedly shared by multiple threads of different GPUs. This will cause undefined behavior for `activation_fix`, and batch norm buffer update for `running_mean` and `running_var`, as different GPU is doing this according to different split of data, and concurrently write/read the config variables. (They also do not exist on the same devices...)

If the config is registered as buffer, this issue will be solved. @Gaffey 

And, to keep the multi-gpu training as close as possible to single-gpu training, maybe [https://hangzhang.org/PyTorch-Encoding/nn.html#encoding.nn.SyncBatchNorm](synchronized batch norm) should  be used, also synchronized quantization of activation should be used? (Weights can be quantized separately, as they are the same and not in-place updated on any gpus)

**Comment by walkerning** [comment]
I think it's fixed now by 05d86788e3c8f4
